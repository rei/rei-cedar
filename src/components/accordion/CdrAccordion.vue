<script>
import {
  defineComponent, useCssModule, computed, watch, onMounted, ref, inject,
} from 'vue';
import IconCaretDown from '../icon/comps/caret-down.vue';
import { modifyClassName } from '../../utils/buildClass';
import mapClasses from '../../utils/mapClasses';
/** Vertically-stacked list that allows users to expand and collapse additional content */
export default defineComponent({
  name: 'CdrAccordion',
  components: {
    IconCaretDown,
  },
  props: {
    /**
     * The unique id of an accordion.
     * @demoIgnore true
     */
    id: {
      type: String,
      required: true,
    },
    /**
     * Toggle this value to open/close the accordion.
     * @demoIgnore true
     */
    opened: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a compact style.
     */
    compact: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a border-aligned style.
     */
    borderAligned: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets the heading level
     * @demoIgnore true
     */
    level: {
      type: [String, Number],
      required: true,
    },
    /**
     * Toggles content spacing (padding)
     */
    contentSpacing: {
      type: Boolean,
      default: true,
    },
    /**
     * Sets the readable text on CdrAccordion button (also can be slotted)
     * @demoIgnore true
     */
    label: {
      type: String,
    },
  },
  emits: {
    /**
    * Emits on accordion open/close
    */
    'accordion-toggle': null,
  },

  setup(props, ctx) {
    const unwrap = inject('unwrap', false);
    const isGrouped = inject('grouped', false) ? 'li' : 'div';
    const style = useCssModule();
    const accordionContentEl = ref(null);
    const focused = ref(false);
    const maxHeight = ref(props.opened
      ? 'none'
      : '0px');
    const headingTag = `h${props.level}`;
    const labelClass = 'cdr-accordion__label';
    const baseClass = 'cdr-accordion';
    const iconClass = 'cdr-accordion__icon';
    const containerClass = 'cdr-accordion__content-container';
    const contentClass = 'cdr-accordion__content';
    const isUnwrapped = computed(() => unwrap && unwrap.isUnwrapped);
    const headingContent = computed(() => (isUnwrapped.value
      ? 'div'
      : 'button'));
    const headingContentStyle = computed(() => (isUnwrapped.value
      ? 'js-cdr-accordion-button'
      : [style['cdr-accordion__button'], 'js-cdr-accordion-button']));
    const headingClass = computed(() => (isUnwrapped.value
      ? 'cdr-accordion__header--unwrapped'
      : 'cdr-accordion__header'));
    const compactClass = computed(() => (props.compact
      ? modifyClassName(baseClass, 'compact')
      : ''));
    const borderAlignedClass = computed(() => (props.borderAligned
      ? modifyClassName(baseClass, 'border-aligned')
      : ''));
    const focusedClass = computed(() => (focused.value
      ? modifyClassName(baseClass, 'focused')
      : null));
    const unwrapClass = computed(() => (isUnwrapped.value
      ? modifyClassName(baseClass, 'unwrap')
      : null));
    const noSpacingClass = computed(() => (!props.contentSpacing
      ? modifyClassName(baseClass, 'no-spacing')
      : null));

    const isOpenClass = computed(() => (props.opened || isUnwrapped.value
      ? 'cdr-accordion--open'
      : 'cdr-accordion--closed'));

    const onClick = (event) => {
      ctx.emit('accordion-toggle', event);
    };
    const onFocus = () => {
      focused.value = true;
    };
    const onBlur = () => {
      focused.value = false;
    };

    const listeners = computed(() => (isUnwrapped.value
      ? {}
      : {
        click: onClick,
        focus: onFocus,
        blur: onBlur,
      }));
    watch(() => props.opened, (opened) => {
      maxHeight.value = !opened ? `${accordionContentEl.value.clientHeight}px` : '0px';
      // nextTick is not sufficient here, must wait for CSS to re-paint
      setTimeout(() => {
        // on next frame, set maxHeight to new value
        maxHeight.value = opened ? `${accordionContentEl.value.clientHeight}px` : '0px';
        setTimeout(() => {
          // after animation is complete, remove max-height so content can reflow
          maxHeight.value = opened ? 'none' : '0px';
        }, 350); // cdr-duration-3x + 50ms
      }, 50);
    });

    onMounted(() => {
      if (props.opened && accordionContentEl.value) {
        maxHeight.value = 'none';
      }
    });
    return {
      style: useCssModule(),
      headingTag,
      headingClass,
      headingContent,
      headingContentStyle,
      isGrouped,
      unwrap,
      accordionContentEl,
      focused,
      maxHeight,
      baseClass,
      labelClass,
      compactClass,
      borderAlignedClass,
      focusedClass,
      iconClass,
      containerClass,
      isOpenClass,
      contentClass,
      noSpacingClass,
      listeners,
      unwrapClass,
      isUnwrapped,
      mapClasses,
    };
  },
});

</script>

<template>
  <component
    :is="isGrouped"
    :class="!isUnwrapped
      ? mapClasses(style, baseClass, compactClass, borderAlignedClass, focusedClass, noSpacingClass)
      : null"
    :id="`${id}-accordion`"
  >
    <component
      :is="headingTag"
      :class="style[headingClass]"
    >
      <!--
        Triggered on accordion open/close
        @event accordion-toggle
      -->
      <component
        :is="headingContent"
        :class="headingContentStyle"
        :id="id"
        v-on="listeners"
        :aria-expanded="!isUnwrapped ? `${opened}` : null"
        :aria-pressed="!isUnwrapped ? `${opened}` : null"
        :aria-controls="!isUnwrapped ? `${id}-collapsible` : null"
      >
        <span
          :class="style[labelClass]"
          :id="`${id}-label`"
        >
          <!-- @slot  Sets the readable text on the CdrAccordion button -->
          <slot name="label">
            {{ label }}
          </slot>
        </span>
        <icon-caret-down
          v-if="!isUnwrapped"
          :class="mapClasses(style, iconClass, isOpenClass)"
          :size="compact ? 'small' : null"
        />
      </component>
    </component>
    <div
      :class="mapClasses(style, containerClass, isOpenClass, unwrapClass)"
      :style="{ maxHeight: isUnwrapped ? 'none' : maxHeight }"
    >
      <div
        :class="mapClasses(style, contentClass, isOpenClass, unwrapClass)"
        :aria-hidden="!isUnwrapped ? `${!opened}` : null"
        :id="`${id}-collapsible`"
        ref="accordionContentEl"
      >
        <!-- @slot CdrAccordion content -->
        <slot />
      </div>
    </div>
  </component>
</template>

<style lang="scss" module src="./styles/CdrAccordion.module.scss">
</style>
