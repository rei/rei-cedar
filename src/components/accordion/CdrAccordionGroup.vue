<script>
import {
  defineComponent, useCssModule, computed, ref, reactive, onMounted, provide,
} from 'vue';
import { debounce } from 'lodash-es';
import propValidator from '../../utils/propValidator';
import getCurrentBreakpoint from '../../mixins/breakpoints';

export default defineComponent({
  name: 'CdrAccordionGroup',
  props: {
    unwrap: {
      type: [String, Boolean],
      default: false,
      validator: (value) => {
        if (typeof value === 'string') {
          return propValidator(
            value,
            ['@xs', '@sm', '@md', '@lg'],
            false,
          );
        }
        return typeof value === 'boolean';
      },
    },
  },

  setup(props) {
    const baseClass = 'cdr-accordion-group';
    const currentIdx = ref(0);
    const accordionButtons = ref([]);
    const accordionGroupEl = ref(null);
    const unwrapped = reactive({
      isUnwrapped: !!props.unwrap,
    });
    provide('unwrap', unwrapped);
    provide('grouped', true);

    const nextIdx = computed(() => {
      const idx = currentIdx.value + 1;
      return idx >= accordionButtons.value.length ? 0 : idx;
    });
    const prevIdx = computed(() => {
      const idx = currentIdx.value - 1;
      return idx <= -1 ? accordionButtons.value.length - 1 : idx;
    });
    const handleKeyDown = (e) => {
      // something besides the button is focused
      if (currentIdx.value === -1) return;

      const { key } = e;
      switch (key) {
        case 'Home':
          e.preventDefault();
          accordionButtons.value[0].focus();
          break;
        case 'End':
          e.preventDefault();
          accordionButtons.value[accordionButtons.value.length - 1].focus();
          break;
        case 'ArrowDown':
        case 'Down':
          e.preventDefault();
          accordionButtons.value[nextIdx.value].focus();
          break;
        case 'ArrowUp':
        case 'Up':
          e.preventDefault();
          accordionButtons.value[prevIdx.value].focus();
          break;
        default: break;
      }
    };
    const focusin = (e) => {
      // find out which, if any, button is focused
      currentIdx.value = Array.prototype.indexOf.call(accordionButtons.value, e.target);
    };

    onMounted(() => {
      accordionButtons.value = accordionGroupEl.value.querySelectorAll('.js-cdr-accordion-button');
      if (typeof props.unwrap === 'string') {
        unwrapped.isUnwrapped = props.unwrap.indexOf(getCurrentBreakpoint()) !== -1;
        window.addEventListener('resize', debounce(() => {
          unwrapped.isUnwrapped = props.unwrap.indexOf(getCurrentBreakpoint()) !== -1;
        }, 300));
      }
    });

    return {
      style: useCssModule(),
      handleKeyDown,
      focusin,
      baseClass,
      accordionGroupEl,
    };
  },
});

</script>

<template>
  <ul
    :class="style[baseClass]"
    ref="accordionGroupEl"
    @focusin="focusin"
    @keydown="handleKeyDown"
  >
    <slot />
  </ul>
</template>

<style lang="scss" module src="./styles/CdrAccordionGroup.module.scss">
</style>
