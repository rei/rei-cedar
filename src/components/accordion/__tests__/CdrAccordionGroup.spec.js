import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrAccordionGroup from '../CdrAccordionGroup.vue';
import CdrAccordion from '../CdrAccordion.vue';
import { h } from '@vue/runtime-core';

describe('CdrAccordionGroup', () => {
  let wrapper;
  let buttons;
  describe('when wrapped', () => {
    beforeEach(() => {
      wrapper = mount(CdrAccordionGroup, {
        slots: {
          default: [
            h(CdrAccordion, { id: 'tab1', level: '2', label: 'label1' }),
            h(CdrAccordion, { id: 'tab2', level: '2', label: 'label2' }),
            h(CdrAccordion, { id: 'tab3', level: '2', label: 'label3' }),
          ],
        },
        attachTo: document.body
      });
      buttons = wrapper.findAll('.js-cdr-accordion-button');
      buttons[0].element.focus();
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    });

    describe('keyboard interactions', ()=>{
      it('hitting ArrowUp focuses on the expected element', async () => {
        buttons[0].element.focus();
        wrapper.trigger('keydown', { key: 'ArrowUp' });
        await wrapper.vm.$nextTick();
        expect(buttons[2].element).toBe(document.activeElement);
      });

      it('hitting ArrowDown focuses on the expected element', async () => {
        buttons[2].element.focus();
        wrapper.trigger('keydown', { key: 'ArrowDown' });
        await wrapper.vm.$nextTick();
        expect(buttons[0].element).toBe(document.activeElement);
      });

      it('hitting End focuses on the expected element', async () => {
        wrapper.trigger('keydown', { key: 'End' });
        await wrapper.vm.$nextTick();
        expect(buttons[2].element).toBe(document.activeElement);
      });

      it('hitting Home focuses on the expected element', async () => {
        wrapper.trigger('keydown', { key: 'Home' });
        await wrapper.vm.$nextTick();
        expect(buttons[0].element).toBe(document.activeElement);
      });
    })
  });

  describe('when unwrapped', () => {
    let wrapper;
    beforeEach(async () => {
      wrapper = mount(h(CdrAccordionGroup, { unwrap: true }, {
        default: () => [
          h(CdrAccordion, { id: 'tab1', level: '2', label: 'label1' }),
          h(CdrAccordion, { id: 'tab2', level: '2', label: 'label2' }),
          h(CdrAccordion, { id: 'tab3', level: '2', label: 'label3' }),
        ]
      }));
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    });
  });

  describe('with string value for unwrapped prop', ()=>{
    beforeEach(async () => {
      wrapper = mount(h(CdrAccordionGroup, { unwrap: '@md' }, {
        default: () => [
          h(CdrAccordion, { id: 'tab1', level: '2', label: 'label1' }),
          h(CdrAccordion, { id: 'tab2', level: '2', label: 'label2' }),
          h(CdrAccordion, { id: 'tab3', level: '2', label: 'label3' }),
        ]
      }));
    })
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    });
  })
});
