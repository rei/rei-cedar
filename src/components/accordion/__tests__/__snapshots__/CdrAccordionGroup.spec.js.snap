// Vitest Snapshot v1

exports[`CdrAccordionGroup > when mounted > renders correctly 1`] = `
<ul
  class="cdr-accordion-group"
>
  
  <li
    class="cdr-accordion"
    id="tab1-accordion"
  >
    <h2
      class="cdr-accordion__header"
    >
      <button
        aria-controls="tab1-collapsible"
        aria-expanded="false"
        aria-pressed="false"
        class="cdr-accordion__button js-cdr-accordion-button"
        id="tab1"
      >
        <span
          class="cdr-accordion__label"
          id="tab1-label"
        >
          
          label1
          
        </span>
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-accordion__icon cdr-accordion--closed"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <path
            d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 00-1.414-1.414L12 13.586 7.705 9.29a1 1 0 00-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
            role="presentation"
          />
          
          <!--v-if-->
        </svg>
      </button>
    </h2>
    <div
      class="cdr-accordion__content-container cdr-accordion--closed"
      style="max-height: 0px;"
    >
      <div
        aria-hidden="true"
        class="cdr-accordion__content cdr-accordion--closed"
        id="tab1-collapsible"
      >
        
        
      </div>
    </div>
  </li>
  <li
    class="cdr-accordion"
    id="tab2-accordion"
  >
    <h2
      class="cdr-accordion__header"
    >
      <button
        aria-controls="tab2-collapsible"
        aria-expanded="false"
        aria-pressed="false"
        class="cdr-accordion__button js-cdr-accordion-button"
        id="tab2"
      >
        <span
          class="cdr-accordion__label"
          id="tab2-label"
        >
          
          label2
          
        </span>
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-accordion__icon cdr-accordion--closed"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          
          
          
          <path
            d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 00-1.414-1.414L12 13.586 7.705 9.29a1 1 0 00-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
            role="presentation"
          />
          
          <!--v-if-->
        </svg>
      </button>
    </h2>
    <div
      class="cdr-accordion__content-container cdr-accordion--closed"
      style="max-height: 0px;"
    >
      <div
        aria-hidden="true"
        class="cdr-accordion__content cdr-accordion--closed"
        id="tab2-collapsible"
      >
        
        
      </div>
    </div>
  </li>
  
</ul>
`;

exports[`CdrAccordionGroup > when unwrapped > renders correctly 1`] = `
<ul
  class="cdr-accordion-group"
>
  
  <li
    class=""
    id="tab1-accordion"
  >
    <h2
      class="cdr-accordion__header--unwrapped"
    >
      <div
        class="js-cdr-accordion-button"
        id="tab1"
      >
        <span
          class="cdr-accordion__label"
          id="tab1-label"
        >
          
          label1
          
        </span>
        <!--v-if-->
      </div>
    </h2>
    <div
      class="cdr-accordion__content-container cdr-accordion--open cdr-accordion--unwrap"
      style="max-height: none;"
    >
      <div
        class="cdr-accordion__content cdr-accordion--open cdr-accordion--unwrap"
        id="tab1-collapsible"
      >
        
        
      </div>
    </div>
  </li>
  <li
    class=""
    id="tab2-accordion"
  >
    <h2
      class="cdr-accordion__header--unwrapped"
    >
      <div
        class="js-cdr-accordion-button"
        id="tab2"
      >
        <span
          class="cdr-accordion__label"
          id="tab2-label"
        >
          
          label2
          
        </span>
        <!--v-if-->
      </div>
    </h2>
    <div
      class="cdr-accordion__content-container cdr-accordion--open cdr-accordion--unwrap"
      style="max-height: none;"
    >
      <div
        class="cdr-accordion__content cdr-accordion--open cdr-accordion--unwrap"
        id="tab2-collapsible"
      >
        
        
      </div>
    </div>
  </li>
  <li
    class=""
    id="tab3-accordion"
  >
    <h2
      class="cdr-accordion__header--unwrapped"
    >
      <div
        class="js-cdr-accordion-button"
        id="tab3"
      >
        <span
          class="cdr-accordion__label"
          id="tab3-label"
        >
          
          label3
          
        </span>
        <!--v-if-->
      </div>
    </h2>
    <div
      class="cdr-accordion__content-container cdr-accordion--open cdr-accordion--unwrap"
      style="max-height: none;"
    >
      <div
        class="cdr-accordion__content cdr-accordion--open cdr-accordion--unwrap"
        id="tab3-collapsible"
      >
        
        
      </div>
    </div>
  </li>
  
</ul>
`;
